controllers/cloudstackmachinestatechecker_controller.go

```diff
@@ -67,42 +70,45 @@ func (r *CloudStackMachineStateCheckerReconciler) Reconcile(ctx context.Context,
 }
 
 func (r *CloudStackMachineStateCheckerReconciliationRunner) Reconcile() (ctrl.Result, error) {
-	if res, err := r.GetParent(r.ReconciliationSubject, r.CSMachine)(); r.ShouldReturn(res, err) {
-		return res, err
-	}
-
-	if res, err := r.GetParent(r.CSMachine, r.CAPIMachine)(); r.ShouldReturn(res, err) {
-		return res, err
-	}
-
-	if err := r.CSClient.ResolveVMInstanceDetails(r.CSMachine); err != nil && !strings.Contains(strings.ToLower(err.Error()), "no match found") {
-		return r.ReturnWrappedError(err, "failed to resolve VM instance details")
-	}
-
-	csRunning := r.CSMachine.Status.InstanceState == "Running"
-	csTimeInState := r.CSMachine.Status.TimeSinceLastStateChange()
-	capiRunning := r.CAPIMachine.Status.Phase == "Running"
-
-	// capiTimeout indicates that a new VM is running, but it isn't reachable due to a network issue or a misconfiguration.
-	// When this happens, the machine should be deleted or else the cluster won't ever recover.
-	capiTimeout := csRunning && !capiRunning && csTimeInState > 5*time.Minute
-
-	if csRunning && capiRunning {
-		r.ReconciliationSubject.Status.Ready = true
-	} else if !csRunning || capiTimeout {
-		r.Log.Info("CloudStack instance in bad state",
-			"name", r.CSMachine.Name,
-			"instance-id", r.CSMachine.Spec.InstanceID,
-			"cs-state", r.CSMachine.Status.InstanceState,
-			"cs-time-in-state", csTimeInState.String(),
-			"capi-phase", r.CAPIMachine.Status.Phase)
-
-		if err := r.K8sClient.Delete(r.RequestCtx, r.CAPIMachine); err != nil {
-			return r.ReturnWrappedError(err, "failed to delete CAPI machine")
-		}
-	}
-
-	return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+	return r.RunReconciliationStages(
+		r.GetParent(r.ReconciliationSubject, r.CSMachine),
+		r.GetParent(r.CSMachine, r.CAPIMachine),
+		r.CheckPresent(map[string]client.Object{"CloudStackMachine": r.CSMachine, "Machine": r.CAPIMachine}),
+		r.GetObjectByName("placeholder", r.FailureDomain,
+			func() string { return r.CSMachine.Spec.FailureDomainName }),
+		r.CheckPresent(map[string]client.Object{"CloudStackFailureDomain": r.FailureDomain}),
+		r.AsFailureDomainUser(&r.FailureDomain.Spec),
+		func() (ctrl.Result, error) {
+			if err := r.CSClient.ResolveVMInstanceDetails(r.CSMachine); err != nil {
+				if !strings.Contains(strings.ToLower(err.Error()), "no match found") {
+					return r.ReturnWrappedError(err, "failed to resolve VM instance details")
+				}
+			}
+
+			// capiTimeout indicates that a new VM is running, but it isn't reachable.
+			// The cluster may not recover if the machine isn't replaced.
```

Comment:
Do we have an issue tracking the machinestatechecker bug? If so, I'd love to link that here as TODO